File: product_controller.go
package main

import (
	"github.com/gin-gonic/gin"
	jwt "iknowbook.com/app/jwt"
	. "iknowbook.com/app/product"
)

type ProductController struct {
	service ProductService
	group   *gin.RouterGroup
}

func NewProductController(service ProductService, router *gin.Engine) ProductController {
	return ProductController{
		service: service,
		group:   router.Group("api/v1/product"),
	}
}

func (ctr ProductController) Run() {
	ctr.QueryByConditions()
	ctr.QueryByBarcode()
	ctr.QueryNewest()

	ctr.MaxPrice()

	ctr.Insert()
	ctr.Update()
	ctr.Delete()
}

func (ctr ProductController) Insert() {
	ctr.group.POST(
		"/insert",
		func(ctx *gin.Context) {
			jwt.VerifyBearerToken(ctx, ctr.service.Insert)
		},
	)
}

func (ctr ProductController) Update() {
	ctr.group.PUT(
		"/update",
		func(ctx *gin.Context) {
			jwt.VerifyBearerToken(ctx, ctr.service.Update)
		},
	)
}

func (ctr ProductController) Delete() {
	ctr.group.DELETE(
		"/delete",
		func(ctx *gin.Context) {
			jwt.VerifyBearerToken(ctx, ctr.service.Delete)
		},
	)
}

func (ctr ProductController) QueryByConditions() {
	ctr.group.GET(
		"/query/",
		func(ctx *gin.Context) {
			jwt.VerifyBearerToken(ctx, ctr.service.QueryByConditions)
		},
	)
}

func (ctr ProductController) QueryByBarcode() {
	ctr.group.GET(
		"/query/barcode/:barcode",
		func(ctx *gin.Context) {
			jwt.VerifyBearerToken(ctx, ctr.service.QueryByBarcode)
		},
	)
}

func (ctr ProductController) MaxPrice() {
	ctr.group.GET(
		"/maxprice",
		func(ctx *gin.Context) {
			jwt.VerifyBearerToken(ctx, ctr.service.MaxPrice)
		},
	)
}

func (ctr ProductController) QueryNewest() {
	ctr.group.GET(
		"/query/new/:range",
		ctr.service.QueryNewest,
	)
}

File: static/js/product/product_controller.js
import ProductService from './product_service.js'

/**
 *
 *
 * @export
 * @return {*}
 */
export default function ProductController () {
  const service = ProductService()

  const getProductsByConditions = async (conditions) => {
    return service.getProduct(`/api/v1/product/query/?${conditions}`)
  }

  const getProductsByBarcode = async (barcode) => {
    return service.getProduct(`/api/v1/product/query/barcode/${barcode}`)
  }

  const updateProduct = async (success = () => {}) => {
    return service.updateProduct('/api/v1/product/update', success())
  }

  const deleteProduct = async (success = () => {}) => {
    return service.deleteProduct('/api/v1/product/delete', success())
  }

  const addProduct = async (success = () => {}) => {
    return service.addProduct('/api/v1/product/insert', success())
  }

  return {
    getProductsByConditions: (conditions = '') => getProductsByConditions(conditions),
    getProductsByBarcode: (barcode = '') => getProductsByBarcode(barcode),
    updateProduct: (success = () => {}) => updateProduct(success()),
    deleteProduct: (success = () => {}) => deleteProduct(success()),
    addProduct: (success = () => {}) => addProduct(success())
  }
}
